core/src/apps/solana/sign_tx.py

```diff
@@ -0,0 +1,142 @@
+            br_code=ButtonRequestType.Other,
+        )
+
+    await show_instructions(address_n, signer_public_key, transaction)
+
+    signer_address = base58.encode(seed.remove_ed25519_prefix(node.public_key()))
+
+    await show_final_confirmation(
+        address_n,
+        signer_address,
+        transaction.blockhash,
+        calculate_fee(transaction),
+    )
+
+    signature = ed25519.sign(node.private_key(), serialized_tx)
+
+    return SolanaTxSignature(signature=signature)
+
+
+async def show_instructions(
+    signer_path: list[int], signer_public_key: bytes, transaction: Transaction
+) -> None:
+    instructions_count = len(transaction.instructions)
+    for instruction_index, instruction in enumerate(transaction.instructions, 1):
+        if not instruction.is_program_supported:
+            from .ui import show_unsupported_program_confirm
+
+            await show_unsupported_program_confirm(
+                instruction,
+                instructions_count,
+                instruction_index,
+                signer_path,
+                signer_public_key,
+            )
+        elif not instruction.is_instruction_supported:
+            from .ui import show_unsupported_instruction_confirm
+
+            await show_unsupported_instruction_confirm(
+                instruction,
+                instructions_count,
+                instruction_index,
+                signer_path,
+                signer_public_key,
+            )
+        else:
+            from .ui import show_confirm
+
+            await show_confirm(
+                instruction,
+                instructions_count,
+                instruction_index,
+                signer_path,
+                signer_public_key,
+            )
+
+
+def calculate_fee(transaction: Transaction) -> int:
+    from .constants import (
+        ADDRESS_SIG,
+        SOLANA_BASE_FEE_LAMPORTS,
+        SOLANA_COMPUTE_UNIT_LIMIT,
+    )
+    from .transaction.instructions import (
+        COMPUTE_BUDGET_PROGRAM_ID,
+        COMPUTE_BUDGET_PROGRAM_ID_INS_SET_COMPUTE_UNIT_LIMIT,
+        COMPUTE_BUDGET_PROGRAM_ID_INS_SET_COMPUTE_UNIT_PRICE,
+    )
+
+    number_of_signers = 0
+    for address in transaction.addresses:
+        if address[1] == ADDRESS_SIG:
+            number_of_signers += 1
+
+    base_fee = SOLANA_BASE_FEE_LAMPORTS * number_of_signers
+
+    unit_price: int | None = None
+    unit_limit: int = SOLANA_COMPUTE_UNIT_LIMIT
+
+    for instruction in transaction.instructions:
+        if instruction.program_id == COMPUTE_BUDGET_PROGRAM_ID:
+            if (
+                instruction.instruction_id
+                == COMPUTE_BUDGET_PROGRAM_ID_INS_SET_COMPUTE_UNIT_LIMIT
+            ):
+                unit_limit = instruction.units
+            elif (
+                instruction.instruction_id
+                == COMPUTE_BUDGET_PROGRAM_ID_INS_SET_COMPUTE_UNIT_PRICE
+            ):
+                unit_price = instruction.lamports
```

Comment:
According to the Solana guys: [Compute budget instruction must be in the 1st 3 instructions (avoid nonce marker), otherwise ignored](https://github.com/solana-labs/solana/blob/647aa926673e3df4443d8b3d9e3f759e8ca2c44b/sdk/src/compute_budget.rs#L130-L132)
And multiple instances of that instruction _could_ exist but only the first one is respected, and generally it’s the first instructions.


[Changes](https://github.com/trezor/trezor-firmware/commit/c273d4358d4e80e2773fae78043e14c5b74f0b12)

Reply:
if someone places two instructions, and the first sets unit price to 0 and the second to 10000, the fee calculation will be wrong here.
i'm afraid you'll need to set the `unit_limit` and `unit_price` to `None` first, to correctly detect which instruction set it first (contrary to that previous review comment from me)